src/util/WasMoved.hpp

```diff
@@ -0,0 +1,73 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of clio: https://github.com/XRPLF/clio
+    Copyright (c) 2025, the clio developers.
+
+    Permission to use, copy, modify, and distribute this software for any
+    purpose with or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#pragma once
+
+namespace util {
+
+/**
+ * @brief A base-class that can be used to check whether the current instance was moved from
+ */
+class WasMoved {
+    bool wasMoved_ = false;
+
+protected:
+    /**
+     * @brief The function to be used by clients in order to check whether the instance was moved from
+     * @return true if moved from; false otherwise
+     */
+    [[nodiscard]] bool
+    wasMoved() const noexcept
+    {
+        return wasMoved_;
+    }
+
+protected:
+    WasMoved() = default;  // should only be used via inheritance
+
+public:
+    ~WasMoved() = default;
+
+    /**
+     * @brief Move constructor sets the internal flag to true on the moved-from object
+     * @param other the moved-from object
+     */
+    WasMoved(WasMoved&& other)
+    {
+        other.wasMoved_ = true;
+    }
+
+    /**
+     * @brief Move operator sets the internal flag to true on the moved-from object
+     * @param other the moved-from object
+     * @return reference to self
+     */
+    WasMoved&
+    operator=(WasMoved&& other)
+    {
+        other.wasMoved_ = true;
```

Comment:
We may need to check if other is the current object to handle self assignment .
```
WasMoved a; 
a = std::move(a); 
EXPECT_FALSE(a.wasMoved());
```

Reply:
Ok I think we figured out the correct logic. Take a look at the current tests 👍